### PR TITLE
fix(terminal): filter terminal query sequences from captured output

### LINKED ISSUE #2244: Terminal escape code responses leak to stdin, corrupting subsequent input
## Summary

Terminal query responses (such as DSR - Device Status Report) leak to stdin when using the SDK in interactive terminal environments. This causes escape code sequences to appear as garbage in the shell prompt or corrupt subsequent `input()` calls.

## How the Issue Manifests

When CLI tools or the SDK's Rich-based visualizer send terminal queries (e.g., cursor position requests, background color queries), the terminal responds by writing escape sequences to stdin. If these responses are not consumed, they:

1. **Corrupt the next `input()` call** - The escape codes become part of the user's input
2. **Leak to the shell prompt** - After the SDK exits, garbage like `;1R` or `^[[23;1R` appears in the terminal
3. **Accumulate across conversation turns** - Each agent step can add more leaked data

Example of leaked data appearing after script exit:
```
$ python my_agent.py
[agent runs successfully]
$ ;1R              # ← This garbage appeared from leaked escape codes
rgb:30fb/3708/41af # ← Background color response leaked
```

## Potential Impact on SDK Clients

1. **CLI applications using `input()`** - User prompts become corrupted with escape codes
2. **Interactive agents** - Multi-turn conversations accumulate leaked data
3. **Shell integration** - Users see garbage in their terminal after running SDK-based tools
4. **Automation scripts** - Leaked data can interfere with piped commands or shell scripts

### Environment Compatibility

| Environment | DSR Query Leak | Notes |
|-------------|----------------|-------|
| **Linux (OpenHands runtime)** | ✅ Reproducible | DSR queries leak reliably |
| **macOS Terminal.app** | ✅ Reproducible | More queries leak (including OSC 11) |
| **macOS iTerm2** | ⚠️ Expected | Likely affected (not yet verified) |
| **CI/non-TTY** | ❌ Not applicable | No TTY, no terminal queries |

### Minimal Reproduction (No API Key Needed)

This surgical test directly injects terminal queries to demonstrate the leak mechanism:

```python
#!/usr/bin/env python3
"""Minimal reproduction of ANSI escape code leak."""

import os
import select
import sys
import time

def check_stdin() -> bytes:
    """Check for pending data in stdin (non-blocking)."""
    if not sys.stdin.isatty():
        return b""
    try:
        import termios
        old = termios.tcgetattr(sys.stdin)
        new = list(old)
        new[3] &= ~(termios.ICANON | termios.ECHO)
        new[6][termios.VMIN] = 0
        new[6][termios.VTIME] = 0
        termios.tcsetattr(sys.stdin, termios.TCSANOW, new)
        data = b""
        if select.select([sys.stdin], [], [], 0)[0]:
            data = os.read(sys.stdin.fileno(), 4096)
        termios.tcsetattr(sys.stdin, termios.TCSANOW, old)
        return data
    except Exception:
        return b""

def main():
    print("ANSI Escape Code Leak Test")
    print("=" * 40)

    if not sys.stdout.isatty():
        print("ERROR: Run in a real terminal (not CI/piped)")
        sys.exit(1)

    for turn in range(1, 4):
        print(f"\n--- Turn {turn} ---")
        # Send DSR query (cursor position request)
        # Terminal responds with: ESC [ row ; col R
        print("  Sending DSR query (\\x1b[6n)...")
        sys.stdout.write("\x1b[6n")
        sys.stdout.flush()

        time.sleep(0.1)

        leaked = check_stdin()
        if leaked:
            print(f"  LEAK DETECTED: {len(leaked)} bytes: {leaked!r}")
        else:
            print("  No leak (terminal may not respond)")

    print("\n" + "=" * 40)
    print("If you see ';1R' or similar after this script exits, that's the bug.")

if __name__ == "__main__":
    main()
```

**Expected output on affected systems:**
```
--- Turn 1 ---
  Sending DSR query (\x1b[6n)...
  LEAK DETECTED: 7 bytes: b'\x1b[26;1R'

--- Turn 2 ---
  LEAK DETECTED: 7 bytes: b'\x1b[31;1R'
```

### Real-World Reproduction (Requires API Key)

```python
#!/usr/bin/env python3
"""Real-world reproduction using SDK agent."""

import os
import sys

# Requires: pip install openhands-sdk openhands-tools
# Set: LLM_API_KEY, optionally LLM_MODEL and LLM_BASE_URL

from openhands.sdk import Agent, Conversation, LLM, Tool
from openhands.tools.terminal import TerminalTool

llm = LLM(
    model=os.environ.get("LLM_MODEL", "claude-sonnet-4-20250514"),
    api_key=os.environ["LLM_API_KEY"],
    base_url=os.environ.get("LLM_BASE_URL"),
)

agent = Agent(llm=llm, tools=[Tool(name=TerminalTool.name)])
conversation = Conversation(agent=agent, workspace="/tmp")

# Commands with spinners (like gh) trigger more queries
conversation.send_message("Run: gh pr list --repo OpenHands/openhands --limit 3")
conversation.run()
conversation.close()

# After exit, watch for garbage in your shell prompt
```

*The full reproduction script with additional diagnostics is available in the [reproduction gist](https://gist.github.com/jpshackelford/d81954bc0291f77b801397461e72b6a8). To run it:*

```bash
git clone https://gist.github.com/d81954bc0291f77b801397461e72b6a8.git escape-leak-repro
cd escape-leak-repro
export LLM_API_KEY=your-api-key
uv run --with openhands-sdk --with openhands-tools python repro_real.py
```

## Root Cause

Terminal queries are sent by:
- **Rich library** (used by SDK visualizer) - for terminal capability detection
- **CLI tools with spinners** (gh, npm, etc.) - for cursor positioning
- **Progress bars and live displays** - for screen updates

The terminal responds asynchronously by writing escape sequences to stdin. Without explicit consumption, these responses accumulate.

## Proposed Solution

Add a `flush_stdin()` function that drains pending stdin data using non-blocking reads. Call it:
1. After each `agent.step()` in `LocalConversation.run()`
2. Before rendering in `DefaultConversationVisualizer.on_event()`
3. At process exit via `atexit`

See the proposed patch in the [original gist](https://gist.github.com/jpshackelford/d81954bc0291f77b801397461e72b6a8).

## Related

- Original issue: [jpshackelford/lxa#7](https://github.com/jpshackelford/lxa/issues/7)
- Reproduction gist: https://gist.github.com/jpshackelford/d81954bc0291f77b801397461e72b6a8

### PR DESCRIPTION
**Pull request by jpshackelford**
## Summary

Fixes #2244 - Filter terminal query sequences from captured PTY output to prevent visible escape code garbage.

## Problem

When CLI tools like `gh` run inside the SDK's PTY, they send terminal query sequences (DSR, OSC 11, etc.) as part of their progress/spinner UI. These queries get captured as output and when displayed, the terminal processes them and responds, causing visible garbage like:

```
^[[38;1R^[]11;rgb:30fb/3708/41af^G
```

## Root Cause Analysis

The diagnostic script `.pr/diagnose_source.py` confirmed the queries are **IN the captured PTY output**, not generated by terminal responses to Rich rendering. The `gh` command specifically writes:
- `\x1b[6n` (DSR - cursor position query)
- `\x1b]11;?` (OSC 11 - background color query)

## Solution

Added `filter_terminal_queries()` in the terminal tool to strip query sequences that trigger terminal responses while preserving legitimate formatting codes (colors, bold, etc.):

- **Filtered**: DSR, OSC 10/11/4, DA, DA2, DECRQSS
- **Preserved**: ANSI colors, cursor movement, bold/formatting

Applied in `_get_command_output()` before returning to the visualizer.

## Manual Testing

Verified with `uv run python .pr/test_real_world.py`:
- ✅ No visible escape codes in output
- ✅ Colors from `gh` preserved
- ✅ Clean shell prompt after exit

## Files Changed

- `openhands-tools/.../utils/escape_filter.py` (NEW) - Filter implementation
- `openhands-tools/.../terminal_session.py` - Apply filter in output processing
- `tests/tools/terminal/test_escape_filter.py` (NEW) - 16 tests
- `.pr/` - Diagnostic scripts for reviewers

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.13-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.13-nodejs22) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:edca942-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-edca942-python \
  ghcr.io/openhands/agent-server:edca942-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:edca942-golang-amd64
ghcr.io/openhands/agent-server:edca942-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:edca942-golang-arm64
ghcr.io/openhands/agent-server:edca942-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:edca942-java-amd64
ghcr.io/openhands/agent-server:edca942-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:edca942-java-arm64
ghcr.io/openhands/agent-server:edca942-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:edca942-python-amd64
ghcr.io/openhands/agent-server:edca942-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-amd64
ghcr.io/openhands/agent-server:edca942-python-arm64
ghcr.io/openhands/agent-server:edca942-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-arm64
ghcr.io/openhands/agent-server:edca942-golang
ghcr.io/openhands/agent-server:edca942-java
ghcr.io/openhands/agent-server:edca942-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `edca942-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `edca942-python-amd64`) are also available if needed

### CODE REVIEW

**Comment by jpshackelford** [comment]
## Note on API Breakage Check Comment

The API breakage comment above shows "Failed" but this is **misleading and not blocking**.

### What's happening:
- The Griffe check compares this PR against the **last released version** (v1.11.4), not main
- PR #2243 merged a `server_image` default change to main, but this change is not yet in a release
- So Griffe sees a "breaking change" when comparing against v1.11.4, even though this PR doesn't introduce that change

### Why it's not blocking:
- The actual CI check shows ✅ **passing** (see PR checks above)
- The workflow has `continue-on-error: true` for non-release PRs
- This PR was rebased onto main and no longer contains the `dev_workspace.py` change

### Resolution:
The comment will auto-correct once a new version is released that includes the `server_image` change from PR #2243.

File: tests/sdk/logger/test_flush_stdin.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        with mock.patch.object(sys.stdin, "isatty", return_value=True):
+            with mock.patch.dict(sys.modules, {"termios": None}):
+                # Force ImportError by making termios module fail to import
+                with mock.patch(
+                    "openhands.sdk.logger.logger.flush_stdin"
+                ) as mock_flush:
+                    # Since we can't easily mock ImportError for termios,
+                    # we test that the function handles it gracefully
+                    mock_flush.return_value = 0
+                    result = mock_flush()
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply tests/sdk/logger/test_flush_stdin.py]: Fixed in 012422c2. Replaced the useless mock-the-function-being-tested pattern with a proper skip that documents Windows CI covers this path. On Unix platforms, the test now skips with a clear explanation.

File: openhands-workspace/openhands/workspace/docker/dev_workspace.py [hunk 2]
```diff
@@ -28,6 +28,13 @@ class DockerDevWorkspace(DockerWorkspace):
             result = workspace.execute_command("ls -la")
     """
 
+    # Override parent's server_image default to None so that callers
+    # providing base_image don't need to explicitly pass server_image=None.
+    server_image: str | None = Field(
+        default=None,
+        description="Pre-built agent server image. Mutually exclusive with base_image.",
+    )
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply openhands-workspace/openhands/workspace/docker/dev_workspace.py:36]: Fixed by rebasing onto main (8730c481). This change was already merged to main in PR #2243, so it was a duplicate. The PR no longer modifies `dev_workspace.py`.

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py [hunk 3]
```diff
@@ -621,6 +621,12 @@ def run(self) -> None:
                     )
                     iteration += 1
 
+                    # Flush any pending terminal query responses that may have
+                    # accumulated during the step (Rich display, tool execution, etc.)
+                    # This prevents ANSI escape codes from leaking to stdin.
+                    # See: https://github.com/OpenHands/software-agent-sdk/issues/2244
+                    flush_stdin()
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py:665]: The double flush is intentional and serves different purposes:

1. **Visualizer flush (before render)**: Clears escape codes before Rich sends more terminal queries during rendering
2. **Run loop flush (after step)**: Catches anything that accumulated during the step, and is essential when `visualizer=None`

The visualizer is optional (`visualizer: ... | None = DefaultConversationVisualizer`), so we need the run loop flush for the no-visualizer case.

Flushing is idempotent and cheap (returns immediately if nothing to flush), so the defensive double-flush doesn't hurt. Keeping both for robustness.

File: tests/sdk/logger/test_flush_stdin.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        """flush_stdin should be imported in LocalConversation."""
+        from openhands.sdk.conversation.impl import local_conversation
+
+        assert hasattr(local_conversation, "flush_stdin")
+
+    def test_flush_stdin_imported_in_default_visualizer(self):
+        """flush_stdin should be imported in DefaultConversationVisualizer."""
+        from openhands.sdk.conversation.visualizer import default
+
+        assert hasattr(default, "flush_stdin")
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply tests/sdk/logger/test_flush_stdin.py]: Fixed in 012422c2. Replaced the import-only tests with actual behavioral tests:

1. `test_flush_stdin_called_in_visualizer_on_event()` - Verifies flush_stdin is called when visualizer processes an event
2. `test_flush_stdin_registered_with_atexit()` - Verifies the atexit handler is registered

These now test real integration points rather than just checking imports exist.

File: openhands-sdk/openhands/sdk/logger/logger.py [hunk 5]
```diff
@@ -193,3 +195,72 @@ def get_logger(name: str) -> logging.Logger:
...
+            try:
+                termios.tcsetattr(_sys.stdin, termios.TCSANOW, old)
+            except (OSError, termios.error):
+                pass
+    return flushed
+
+
+# Register cleanup at module load time
+if not _cleanup_registered:
+    atexit.register(flush_stdin)
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply openhands-sdk/openhands/sdk/logger/logger.py:451]: Thank you for the positive feedback! 🙏

**Comment by jpshackelford** [comment]
@juanmichelini This PR doesn't change prompts but it does impact the way OpenHands manipulates shell commands and tools and I wouldn't want it to have any kind of negative impacts on our benchmarks. Not sure if you have a process for watching higher risk PRs or doing light weight evals on them. You'd definitely have a better eye than me for what could create issues...

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (enyst) [reply openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py:665]: If visualizer is None, do we even have a rendering problem? I mean, we don't render 😅

**Comment by enyst** [comment]
@openhands Understand this PR and the linked issue it aims to solve, assess the issue yourself. Do a /codereview-roasted on this PR.

**Comment by enyst** [comment]
> Not sure if you have a process for watching higher risk PRs or doing light weight evals on them. 

We typically figure it out at review time, and yes, we used to run some ~50 instances evals on them. This is a good point... the reviewer agent doesn't know.

@openhands Read this https://github.com/OpenHands/software-agent-sdk/pull/2245#issuecomment-3977612873 and look up the workflow running the pr-review agent. Understand what prompt file the reviewer uses. Investigate the agent loop in the SDK, and add to the reviewer prompt to not approve PRs that change behavior in a way that risks evaluation performance, leave those to human maintainers.

**Comment by jpshackelford** [comment]
@enyst You must be using a different model for your roasted review : )

**Comment by enyst** [comment]
> @enyst You must be using a different model for your roasted review : )

Yes! That's why I do it! GPT-5.2, as safety net for over-happy Claudes. Granted, lately it probably exaggerates in the other direction, but that's why we haz a use for humans. 😅

**Comment by jpshackelford** [comment]
@enyst Give me a signal when you are done looking at it and I'll address the roasted issues and any others.

**Comment by enyst** [comment]
> @enyst Give me a signal when you are done looking at it and I'll address the roasted issues and any others.

All good, the agent took the proposed reviewer agent changes to another PR.

Re: the roasted review. If you want, we could also reuse that conversation to ask it for a re-assessment of its own roasted review, sometimes it figures out where it exaggerated (or not).

Actually let me try it

**Comment by enyst** [comment]
(OpenHands-GPT-5.2; note that in general, it's so prompt-sensitive that once I used words like "re-assessment" it will strongly push it to give a different answer/framing; still the core stuff should(tm) be apparent)

---

### Re-evaluation of my review (section by section)

#### Taste rating
After thinking harder: **still “needs improvement,” but closer to “yellow/acceptable with a couple fixes” than “red dumpster fire.”** The core idea is fine; the main question is whether the implementation is *safe* and *provably correct*.

---

## [CRITICAL ISSUES] — are they really critical?

### 1) Termios restore bug due to shallow copy  
**Claim:** `[logger.py L241–259]` `new = list(old)` shares `old[6]`, so mutating `new[6][VMIN/VTIME]` mutates `old` too, so the restore doesn’t restore.

**Is it really the case?** Yes. `termios.tcgetattr()` returns a list where `cc` is a nested list/array at index 6. A shallow copy (`list(old)` / `old.copy()`) will still share that nested `cc`. So `new[6][...] = ...` mutates `old[6]`.

**Is it as major as I made it sound?** Slightly nuanced:
- In the *common* case (terminal in canonical mode, `ICANON` on), `VMIN/VTIME` aren’t used, so leaving them at 0 typically won’t break basic `input()` behavior.
- But in *non-canonical* situations (prompt_toolkit, curses-ish apps, anything that already turned off `ICANON`), you’ve now silently altered the terminal’s read behavior by “restoring” mutated settings. That’s the kind of thing that creates “random” terminal weirdness.

**Conclusion:** Still a real correctness bug and still worth treating as **must-fix**, mostly because it’s trivial to fix and avoids terminal-state corruption in edge-but-real environments.

---

### 2) “flush before render can drop real user input”
**Claim:** `[visualizer default.py L252–257]` flushing stdin before rendering can discard typeahead / real keystrokes.

**Is it really the case?** Yes. The function drains *whatever bytes are pending* on stdin. If the user types while output is happening (typeahead), those bytes are indistinguishable from terminal query responses at this level and will be eaten.

**Is it as major as I made it sound?** This one depends on product expectations:
- If the SDK’s intended UX is “agent runs, user waits, then user types when prompted,” then losing typeahead is less likely, and some projects accept that tradeoff.
- If users *do* type ahead (or you embed this in an app that does concurrent reading, or you’re in a rich TUI-ish environment), silently discarding input is a real UX regression.

Also: calling `flush_stdin()` **before every visualizer render** increases the odds of catching typeahead compared to calling it at a single “safe boundary” (like immediately before prompting).

**Conclusion:** The risk is real; I’d downgrade it from “definitely critical everywhere” to **high-risk behavior change** that needs justification or a narrower call site. The “right” fix depends on intended interaction model.

---

## [IMPROVEMENT OPPORTUNITIES] — did I overstate them?

### 3) Import-time `atexit` side effect  
**Is it really bad?** It’s a side effect, but lots of libraries do import-time setup. Here, the SDK already does import-time things (e.g., logging auto-config). So this is **not catastrophic**.

**Severity after reflection:** **Minor-to-moderate**. Worth noting, but not a blocker by itself.

### 4) Visualizer importing `openhands.sdk.logger` pulls in `litellm` etc.  
**Is it really the case?** Yes: `openhands.sdk.logger.__init__` imports from `.logger`, which imports `litellm`.

**Is it major?** Probably **minor** in this repo because `litellm` is already a core dependency for the SDK runtime. The cleanliness argument (“flush_stdin should live in a lightweight tty util”) still stands, but it’s not a correctness issue.

---

## [TESTING GAPS] — did I demand too much?

### 5) Tests don’t prove actual draining / don’t catch the shallow-copy restore bug  
**Is it really the case?** Yes:
- The tests validate early returns and that calls happen.
- They don’t validate “bytes were present and got drained,” nor do they validate that termios settings are restored correctly (which would catch the shallow-copy issue).

**Is this blocking?** I’d say **not strictly blocking** if maintainers are OK relying on manual reproduction + the small surface area of the function. But for a terminal/tty fix (notoriously environment-sensitive), a single robust PTY-based test would materially increase confidence.

**Severity after reflection:** **Moderate** (nice to have, but I wouldn’t hold up merging solely on this if the core code is fixed).

---

## Bottom line (after deeper thought)
- The **termios shallow-copy/restore** issue is still the one I’m most confident is *objectively wrong* and worth fixing.
- The **“flush on every render”** concern is real, but its severity depends on whether the SDK promises to preserve typeahead / interactive keystrokes during agent output. I still think it’s the highest “behavior change” risk in the PR.
- The other points (atexit registration, heavy import) are **real but secondary**.
- The tests are **thin** for the main behavior, but that’s more about confidence than correctness.

If you want, I can also sanity-check *where and when* user input is actually read in the SDK flow to better judge whether “typeahead loss” is a real practical regression for typical clients, or mostly theoretical.

**Comment by enyst** [comment]
(HUMAN) Last note:
> If users do type ahead (or you embed this in an app that does concurrent reading, or you’re in a rich TUI-ish environment), silently discarding input is a real UX regression.

We support user messages sent at any time, including during agent's run, for sure.

All yours!

**Comment by jpshackelford** [comment]
## Manual Testing Results on macOS

### Testing Approach

Tested this PR branch on macOS using the reproduction scripts from the [linked gist](https://gist.github.com/jpshackelford/d81954bc0291f77b801397461e72b6a8).

**Installation command:**
```bash
uv run --with "openhands-sdk @ git+https://github.com/OpenHands/software-agent-sdk.git@fix/stdin-escape-code-leak#subdirectory=openhands-sdk" \
       --with openhands-tools python repro_real.py
```

### Results

| Test Criteria | Result |
|--------------|--------|
| Garbage in shell prompt after exit | ✅ **None observed** |
| `!!! LEAK DETECTED` message | ✅ **Not triggered** |
| Escape codes visible inline during execution | ⚠️ **Still present** |

**Observed inline escape codes:**
```
^[]11;rgb:30fb/3708/41af^G^[[57;1R
```

These are:
- OSC 11 responses (terminal background color)
- DSR cursor position responses

### Assessment

The fix **successfully addresses the main bug**:
- ✅ No stdin corruption affecting subsequent `input()` calls
- ✅ No garbage characters leaking to shell prompt after script exits
- ✅ `flush_stdin()` is consuming escape code responses before they can leak

**Minor cosmetic issue remaining:** Escape codes still appear inline in the terminal output between the Observation and Agent Action rendering. This appears to be a timing issue where terminal responses are echoed to stdout before `flush_stdin()` consumes them.

### Next Steps

1. **This PR can be merged** - it fixes the critical stdin leak bug as designed
2. **Consider follow-up enhancement:** Add `flush_stdin()` calls at additional points to prevent inline escape code visibility
3. The inline escape codes are cosmetic and do not affect functionality

**Comment by jpshackelford** [comment]
## Manual Testing Results on macOS

### Testing Approach

Tested this PR branch on macOS using the reproduction scripts from the [linked gist](https://gist.github.com/jpshackelford/d81954bc0291f77b801397461e72b6a8).

**Installation command:**
```bash
uv run --with "openhands-sdk @ git+https://github.com/OpenHands/software-agent-sdk.git@fix/stdin-escape-code-leak#subdirectory=openhands-sdk" \
       --with openhands-tools python repro_real.py
```

### Results

| Test Criteria | Result |
|--------------|--------|
| Garbage in shell prompt after exit | ✅ **None observed** |
| `!!! LEAK DETECTED` message | ✅ **Not triggered** |
| Escape codes visible inline during execution | ⚠️ **Still present** |

**Observed inline escape codes:**
```
^[]11;rgb:30fb/3708/41af^G^[[57;1R
```

These are:
- OSC 11 responses (terminal background color)
- DSR cursor position responses

### Assessment

The fix successfully addresses the main stdin leak bug:
- ✅ No stdin corruption affecting subsequent `input()` calls
- ✅ No garbage characters leaking to shell prompt after script exits

However, escape codes still appear inline in the terminal output between the Observation and Agent Action rendering.

### Suggested Next Steps

The inline escape codes appear to be a timing issue - terminal responses are being echoed to stdout before `flush_stdin()` can consume them. To address this:

1. **Identify the source of terminal queries** - The OSC 11 (background color) and DSR queries are likely coming from Rich's terminal capability detection. Adding logging or breakpoints around `flush_stdin()` calls could help pinpoint when queries are sent vs when responses arrive.

2. **Consider adding `flush_stdin()` before Rich renders** - Currently the PR calls `flush_stdin()` after agent steps and before rendering in the visualizer. It may help to also call it immediately before Rich prints output, to catch any responses that arrived during processing.

3. **Alternative: suppress terminal queries** - Rich can be configured to skip terminal capability detection. Setting `TERM=dumb` or using Rich's `force_terminal=False` might prevent the queries entirely, though this could affect output formatting.

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (enyst) [reply openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py:665]: OpenHands, when you review this, read also https://github.com/OpenHands/software-agent-sdk/pull/2245#issuecomment-3977627820

**Comment by jpshackelford** [comment]
## Fix for termios shallow-copy bug

Based on the [detailed analysis](https://github.com/OpenHands/software-agent-sdk/pull/2245#issuecomment-3977652275), I've pushed a fix for the termios shallow-copy issue identified in the review.

### The Problem

The original code used `list(old)` which creates a shallow copy:

```python
old = termios.tcgetattr(_sys.stdin)
new = list(old)  # ← Shallow copy
new[6][termios.VMIN] = 0  # ← Also modifies old[6]!
```

Since `old[6]` (the cc array) is itself a list, both `old[6]` and `new[6]` pointed to the **same object**. Modifying `new[6][VMIN]` also corrupted `old[6]`, making the restore at the end ineffective.

Verified with a quick test:
```python
>>> old[6] is new[6]
True  # Same object!
>>> new[6][termios.VMIN] = 99
>>> old[6][termios.VMIN]
99  # Oops, old is corrupted too
```

### The Fix (commit ae64c17f)

Changed to a proper deep copy using a comprehension that copies nested lists:

```python
new = [item[:] if isinstance(item, list) else item for item in old]
```

Now `old[6]` and `new[6]` are separate objects, and the termios restore works correctly.

---

This addresses the **"termios shallow-copy/restore"** concern from the review. The **typeahead loss** question is still an open design discussion that needs product input.

**Comment by jpshackelford** [comment]
## Added PTY-based tests (commit a8b4af99)

Added two new tests using pseudo-terminals to verify real terminal behavior:

### 1. `test_flush_stdin_restores_termios_settings`
Verifies that VMIN/VTIME settings in the cc array are properly restored after `flush_stdin()`. **This test would have caught the shallow-copy bug** - if we revert the deep-copy fix, this test fails.

### 2. `test_flush_stdin_drains_pending_data`  
Verifies that `flush_stdin()` actually reads and discards pending escape code data from stdin, returning the correct byte count.

Both tests:
- Use `pty.openpty()` to create a real terminal environment
- Are skipped on Windows where the pty module is unavailable
- Test actual behavior, not mocks

---

**Remaining open item**: The typeahead loss concern still needs a product decision.

**Comment by jpshackelford** [comment]
## Typeahead Concern - Problem & Proposed Direction

**Problem**: `flush_stdin()` drains ALL pending stdin data, which would discard legitimate user input typed during agent execution.

**Constraints**:
- ❌ Cannot accept discarding user typeahead - SDK supports messages sent anytime during agent run
- ❌ Cannot limit fix to only atexit - escape codes leak during execution, not just at exit

**Rejected approaches**:
- Flush everything (current impl) - loses user input
- Only flush at exit - doesn't fix mid-execution leaks
- `TIOCSTI` re-injection - disabled on modern Linux kernels for security

**Proposed direction**: Selective flushing
- Parse pending stdin data
- Discard only recognized escape sequences (CSI `\x1b[...`, OSC `\x1b]...`)  
- Buffer any other data (likely user input)
- Export `get_buffered_input()` for SDK to prepend to next input read

Will prototype this approach after addressing CI failures.

File: openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpshackelford) [reply openhands-sdk/openhands/sdk/conversation/impl/local_conversation.py:665]: This will be revisited as part of the selective flushing implementation. The current double-flush approach may change when we implement escape-code-only filtering with input buffering. Resolving for now.

**Comment by jpshackelford** [comment]
## Selective Flushing Implementation (commit 5c34edb)

Based on the [typeahead concern discussion](https://github.com/OpenHands/software-agent-sdk/pull/2245#issuecomment-3977750749), I've implemented selective flushing that preserves user input while discarding terminal escape sequences.

### How It Works

The new implementation parses stdin data byte-by-byte:

1. **CSI sequences** (`\x1b[...X`) - Identified by looking for the final byte (0x40-0x7E per ECMA-48). Complete sequences are discarded; incomplete ones are preserved.

2. **OSC sequences** (`\x1b]...\x07` or `\x1b]...\x1b\\`) - Terminated by BEL or ST. Complete sequences are discarded; incomplete ones are preserved.

3. **Other data** - All other bytes (including partial sequences) are preserved in a module-level buffer.

### New Public API

```python
from openhands.sdk.logger import get_buffered_input, clear_buffered_input

# Retrieve preserved user input (clears buffer)
buffered = get_buffered_input()

# Explicitly clear the buffer without retrieving
clear_buffered_input()
```

### SDK Integration Point

Where the SDK reads user input (e.g., before `input()` calls), it should prepend any buffered data:

```python
buffered = get_buffered_input()
user_input = buffered.decode('utf-8', errors='replace') + input()
```

### Test Coverage

Added comprehensive tests for:
- CSI final byte detection
- CSI/OSC sequence end finding (complete and incomplete)
- Mixed content parsing (escape sequences + user input)
- Buffer retrieval and clearing
- PTY-based integration tests verifying selective behavior

### Design Notes

- **Incomplete sequences**: If an escape sequence is incomplete at the end of a buffer read, it's preserved rather than discarded. This handles the edge case where a user presses Escape and we catch the data before they finish typing.

- **Arrow keys**: Complete CSI sequences like arrow keys (`\x1b[A`) are flushed. This is acceptable because arrow keys typed during agent execution aren't meaningful input. If this becomes an issue, we could add a whitelist.

- **Buffer size**: No explicit limit on preserved input buffer. In practice, user typeahead during agent execution should be minimal.

**Comment by jpshackelford** [comment]
Next steps, another round of manual testing.

**Comment by jpshackelford** [comment]
## Manual Testing Report

### Test Environment
- macOS 15.6.1
- Local SDK development environment with `uv run`
- Terminal: iTerm2

### Root Cause Discovery

The diagnostic script `.pr/diagnose_source.py` was key to identifying the root cause. Running it captured the raw PTY output:

```
Raw captured output (repr):
'\x1b]11;?\x1b\\\x1b[6n\x1b]11;?\x1b\\\x1b[6n\nShowing 3 of 183 open pull requests...'
```

**Finding**: The `gh` command writes terminal query sequences directly to its stdout as part of its spinner UI:
- `\x1b[6n` - DSR (cursor position query)
- `\x1b]11;?` - OSC 11 (background color query)

These get captured by the PTY and when displayed, the terminal responds to them, causing visible garbage.

### Fix Verification

After implementing the filter, ran `uv run python .pr/test_real_world.py`:

- ✅ **No visible escape codes** in output during conversation
- ✅ **Colors preserved** - `gh` output still has colored PR IDs, branches
- ✅ **Clean shell prompt** after script exits - no leftover garbage

### Previous Attempts (documented in .pr/)

1. **flush_stdin()** - Flushed pending input but didn't help (queries are in stdout, not stdin responses)
2. **Echo suppression** - Disabled terminal echo but didn't help (queries already captured before display)
3. **Filter queries** ✅ - Strips query sequences from captured output before display

The diagnostic scripts in `.pr/` document this investigation for future reference.

**Comment by jpshackelford** [comment]
Next steps. Determine if we can reduce the scope of the code committed PR and still resolve the issue. We may end up closing this PR and opening an alternative if we can simplify. Stay tuned.

**Comment by jpshackelford** [comment]
Next step test and refine https://github.com/OpenHands/software-agent-sdk/pull/2334 as a more targeted solution. If that PR fixes the problem, we will close this one without merging.

**Comment by jpshackelford** [comment]
I believe https://github.com/OpenHands/software-agent-sdk/pull/2334 is the better solution here. @enyst if you wouldn't mind taking a look there, I will close this one.
